### DISCUSSION TITLE
mixed-GPU? Wayland?

### DISCUSSION THREAD

**Discussion by ghost** [YaLTeR/niri#133] [Q&A]
Can someone chime in on what "mixed-GPU" means? Years ago I spoke with a Wayland Dev about multi GPU and basically what I was told was Wayland will dump all the work on to one GPU while turning all the others into worthless output / fb splitters. I had heard Wayland might have been working on making Multi-Gpu work where in each can run their own workloads / screens correctly but news is scarce...

So does Wayland actually run multi/mixed GPU correctly now or is this a niri only thing or does niri do this the way xrandr does it (resulting in high idle loads and horrid performance)? ENLIGHTEN ME PLZ~! (playful yelling...not angry...OK maybe tad scared as Wayland/GTK has been tightening the noose on multi GPU workstations like mine...)

**Comment by YaLTeR** [comment]
Hey!

When I mentioned mixed-GPU in the release notes, what I meant was that if you have a monitor plugged into a different GPU than your primary one, it will still light up and work. This likely works out of the box on wlroots, but on Smithay so far you have to implement it manually.

Now how exactly this is implemented can vary. You absolutely can use different GPUs for rendering different outputs, cosmic-comp even does some intelligent on-the-fly picking of what GPU to render an output with, based on which GPUs the windows visible on that output use.

In niri I went for a simpler option of always rendering all outputs on the primary GPU. Unless the window is fullscreen and gets scanned out directly, then the rendering is skipped altogether.

Another side to this question is which GPUs the apps themselves can use. They can use whatever GPU they want, so by default that would be the primary GPU, or you can switch it with `DRI_PRIME` or `switcherooctl` (same thing as what happens when you press "Launch on a dedicated graphics card" in GNOME Shell). At the moment on Wayland there isn't a perfect way to tell an app what GPU it should use for rendering at runtime though, so you need to do it manually.

**Comment by ghost** [comment]
Hmm sounds like this is still a no go. As soon as someone mentions DRI_PRIME it makes me think they assume some IGP/APU + dedicated laptop setup not a machine full of GPU's all running several screens per GPU like my workstation. There is no GPU switching here, all GPU's are all running dedicated screens all the time. I tell what applications which GPU to run on via DISPLAY=:0.X as each GPU has it's own XScreen allowing task/screen designation. Thanks for the info.

**Reply by YaLTeR** [reply]
That should be possible on cosmic-comp I think. It provides a Wayland socket per GPU that you can use to inform the apps to render there, perhaps together with DRI_PRIME (which as far as I understand tells Mesa which GPU to pick).

**Comment by ghost** [comment]
DRI_PRIME requires you to be using PRIME heh, again all my GPU's are always used, there is no PRIME switching. POP_OS (current/past) does assume and set up my system as a PRIME setup which screws everything up heh. I kinda doubt cosmic-comp will run correctly as well when it's ready. Multi-GPU has become this thing everyone excludes these days. At least when the GPU's are used that is. Everyone assumes multi-gpu means the other GPU's are for Cuda/rendering/ML type workloads and not running screens.

**Reply by YaLTeR** [reply]
No, it's a Mesa variable that can set the exact GPU device to use: https://docs.mesa3d.org/envvars.html#envvar-DRI_PRIME

**Reply by ghost** [reply]
I've read those as a refresher for rebuttles to Steam support non-sense. No one seems to get there is no "PRIME" with my type of setup. GPU's are denoted in my xorg.conf and applications either default to the XScreen I'm on when launched or launched with a specified DISPLAY= prefix. I've written lots of little utilities to control things as more and more projects push the "Wayland" single GPU "future." For example GTK stopped enumerating XScreens about 3.1. This breaks any DE/WM that uses GTK for multi GPU users.

**Reply by YaLTeR** [reply]
Well I'm just saying that as far as I understand how GPUs and Wayland works, you should be able to replicate the same setup on cosmic-comp by setting the right WAYLAND_DISPLAY instead of DISPLAY. I'm not sure what other compositors render screens on their respective GPUs at the moment. In niri it would be quite possible to do the same, just that it needs work and that's a bit of an uncommon requirement.

**Reply by ghost** [reply]
See there is a gap here though but I don't expect you to know, just I was curious if I had missed some news or if niri had some magic going on with Wayland. So I might be able to select a GPU via WAYLAND_DISPLAY but from my talk with the one Wayland dev that still sounds like it won't work correctly. From his explanation the work is dumped on the "primary" GPU and everyone else are just slaves that get passed frame buffer data from the main GPU. Meaning they are super expensive, power sucking extended outputs that output/display but don't actually DO any work. So when you say I could select with WAYLAND_DISPLAY I say OK sure but is that JUST giving me the FB displayed on that GPU's ports or is THAT GPU actually running/rendering? Right now I can have Blender running/rendering on one screen/GPU while another has Unity open in another I'm playing Tarkov while I wait. In the Wayland Model (again from a 2 maybe 3 year old explanation) I'd just set my best GPU on fire trying to do that and none of it would work...I'd SEE the stuff on screen but it would be like using a Pinto to pull a semi trailer heh.

I may fool with a live system that has Wayland but I suspect the original devs explanation holds true. Again I did hear news of what sounded like they might be working on correcting that "useless frame buffer passing" setup so my type of setup will work but who knows when that might be done or if it will actually do what I've done for the last 20ish years.

As a side note to the "Wayland future" AMD drivers also can't be used for my kinda set up anymore...they completely ignore the xconf and will only work with xrandr style configs which obliterates performance unless you have matched GPU's.

**Reply by YaLTeR** [reply]
> From his explanation the work is dumped on the "primary" GPU and everyone else are just slaves that get passed frame buffer data from the main GPU.

Yeah, that's what I mean by "In niri I went for a simpler option of always rendering all outputs on the primary GPU.". That is also how Mutter/GNOME Shell currently works, and how I imagine most compositors currently work.

And by "cosmic-comp even does some intelligent on-the-fly picking of what GPU to render an output with" I mean that cosmic-comp does *not* dump all the work on the "primary" GPU always, so it should work as you want.

> So when you say I could select with WAYLAND_DISPLAY I say OK sure but is that JUST giving me the FB displayed on that GPU's ports or is THAT GPU actually running/rendering?

In cosmic-comp that is intended to select the GPU that the apps actually use for rendering, which in turn will make cosmic-comp render that output with that same GPU (and not "primary").

**Comment by ghost** [comment]
Heh "most compositors currently work." - Most don't work on my setup heh but that's more because GTK. If cosmic-comp does that on Wayland it will be a feat but depending on how close to Gnome the POP_OS DE will be it probably won't be my cup of tea.

At this point I'll be taking Xorg / CTWM to the grave heh. Don't get me wrong, I love CTWM.
